src/device_trezor/device_trezor.cpp

```diff
@@ -0,0 +1,379 @@
+        ::crypto::key_image ki;
+        memcpy(ki.data, buff, 32);
+        memcpy(sig.c.data, buff + 32, 32);
+        memcpy(sig.r.data, buff + 64, 32);
+        ski.push_back(std::make_pair(ki, sig));
+      }
+    }
+
+
+    void device_trezor::tx_sign(wallet_shim * wallet,
+                                const tools::wallet2::unsigned_tx_set & unsigned_tx,
+                                tools::wallet2::signed_tx_set & signed_tx,
+                                hw::tx_aux_data & aux_data)
+    {
+      size_t num_tx = unsigned_tx.txes.size();
+      signed_tx.key_images.clear();
+      signed_tx.key_images.resize(unsigned_tx.transfers.size());
+
+      for(size_t tx_idx = 0; tx_idx < num_tx; ++tx_idx) {
+        std::shared_ptr<protocol::tx::Signer> signer;
+        tx_sign(wallet, unsigned_tx, tx_idx, aux_data, signer);
+
+        auto & cdata = signer->tdata();
+        auto aux_info_cur = signer->store_tx_aux_info();
+        aux_data.tx_device_aux.emplace_back(aux_info_cur);
+
+        // Pending tx reconstruction
+        signed_tx.ptx.emplace_back();
+        auto & cpend = signed_tx.ptx.back();
+        cpend.tx = cdata.tx;
+        cpend.dust = 0;
+        cpend.fee = 0;
+        cpend.dust_added_to_fee = false;
+        cpend.change_dts = cdata.tx_data.change_dts;
+        cpend.selected_transfers = cdata.tx_data.selected_transfers;
+        cpend.key_images = "";
+        cpend.dests = cdata.tx_data.dests;
+        cpend.construction_data = cdata.tx_data;
+
+        std::string key_images;
+        bool all_are_txin_to_key = std::all_of(cdata.tx.vin.begin(), cdata.tx.vin.end(), [&](const cryptonote::txin_v& s_e) -> bool
+        {
+          CHECKED_GET_SPECIFIC_VARIANT(s_e, const cryptonote::txin_to_key, in, false);
+          key_images += boost::to_string(in.k_image) + " ";
+          return true;
+        });
+        if(!all_are_txin_to_key) {
+          throw std::invalid_argument("Not all are txin_to_key");
+        }
+        cpend.key_images = key_images;
+
+        // KI sync
+        size_t num_sources = cdata.tx_data.sources.size();
+        for(size_t src_idx = 0; src_idx < num_sources; ++src_idx){
+          size_t idx_mapped = cdata.source_permutation[src_idx];
+          size_t idx_map_src = cdata.tx_data.selected_transfers[idx_mapped];
+          auto vini = boost::get<cryptonote::txin_to_key>(cdata.tx.vin[src_idx]);
+          signed_tx.key_images[idx_map_src] = vini.k_image;
+        }
+      }
+    }
+
+    void device_trezor::tx_sign(wallet_shim * wallet,
+                   const tools::wallet2::unsigned_tx_set & unsigned_tx,
+                   size_t idx,
+                   hw::tx_aux_data & aux_data,
+                   std::shared_ptr<protocol::tx::Signer> & signer)
+    {
+      AUTO_LOCK_CMD();
+      require_connected();
+      test_ping();
+
+      signer = std::make_shared<protocol::tx::Signer>(wallet, std::addressof(unsigned_tx), idx, std::addressof(aux_data));
```

Comment:
I can substitute `std::addressof` by `&`.

The shared_ptr was added so the `signer` is automatically destroyed after running out of scope (maybe because of an exception) so maybe we can keep it?

Reply:
I'm not asking for a change, just curious about why. vtnerd also does this. It might be simple preference. I'm not asking for the shared_ptr to be removed.

Reply:
Sure I was just not sure whether you mean `std::addressof` or shared_ptr or both.

Reply:
Both the addressof and the make_shared. These are two things I removed in what I wrote :)